### PR TITLE
Expand pytorch input options to iterable schemas

### PR DESCRIPTION
**Pull request by imarquart**
This proposed change eliminates the requirements for the input to a PyTorch model to be a single input/output numpy array, independent of the schema used by the model.

## Related Issues/PRs

See https://github.com/mlflow/mlflow/issues/6820:

### Motivation

> #### What is the use case for this feature?

Mlflow allows considerable flexibility to define a model's signature. Supporting PyTorch is obviously something the Mlflow team strives for (as indicated by the website).

However, the PyTorch wrapper will straight up fail if
1. The input is not a numpy or pandas type - it will even fail if fed a pytorch tensor!
2. An iterable of tensor types
3. Input / Output schemas that would normally be supported by other Mlflow model flavors.

These limitations are not mentioned for other model flavors.

For PyTorch specifically, this is an issue. Increasingly many use cases involve the use of pre-trained models, for instance from HuggingFace.
HuggingFace models specifically take a dict-like input, and thus can not be used.
Similar cases can be made for more complex DL models, for instance

-> Language models with masking tensors
-> Transformer models with memory embeddings of some sort
-> Tabular transformer-models with categorical / continuous tensor inputs (that are embedded differently)
->  Contrastive learning models
-> Retrieval models with paired attention or cross attention
-> Neural search, retrieval enhanced models
-> and many, many more

> #### Why is this use case valuable to support for MLflow users in general?

PyTorch is now the dominant machine learning framework for deep learning models. Mlflow users will want to use deep learning models which typically require more flexible input and output signatures.

> #### Why is this use case valuable to support for your project(s) or organization?

We seek to deploy and train PyTorch models and immediately ran into this issue.

> #### Why is it currently difficult to achieve this use case?

Few SOTA models natively run a single input tensor and a single output tensor. 

## What changes are proposed in this pull request?

Simply expanded the input check for models to 

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

I am unsure about how to test, but I'd recommend just running the current testing suite to see if - as expected - this change does not affect outputs

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Allow PyTorch Model flavor to load inputs and output tensor types in lists, dicts, OrderedDicts etc.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by BenWilson2** [comment]
Looks like you just need to adjust the regex match statement on two existing tests and that portion should be good to go. 
The next thing to validate is to write validation unit tests that predictions can be executed successfully after loading a PyTorch model that's been logged to MLflow with the expanded types. 
Let me know if you have any questions!

**Comment by IngoStatworx** [comment]
> Looks like you just need to adjust the regex match statement on two existing tests and that portion should be good to go.
> The next thing to validate is to write validation unit tests that predictions can be executed successfully after loading a PyTorch model that's been logged to MLflow with the expanded types.
> Let me know if you have any questions!

Hi Ben,

I fixed the regex (I think) so maybe can you trigger a test?

Furthermore, I believe there is some more work to be done on this.
Firstly, I saw that there is a test with transformers (BERT model) which is currently skipped (as transformers library is not installed via test). I might iterate on that.

Second, crucially, I dug around a bit and found that my changes would require a second change, namely in the "Example" class where save_example also tries to coerce the input (ndarray, say) into a Json. Now, coming from list or dict of tensor, this should be straightforward but it still needs to be added for this to actually run.

So it seems even if I add unit tests for the PyTorch Wrapper and make it run, I'll want to actually work on this a bit more in depth to make it a good contrib (let's see weekend).

Do you folks have any tips for setting up the dev environment on M1 Macs?

The default failed for me (pip and docker both).
Maybe conda is the way to go?


Thanks for your help

**Comment by BenWilson2** [comment]
Hi @IngoStatworx we just merged #6850 which adds the transformers package to support recent versions of TF2. Feel free to merge master to your branch. 

For M1 macs, check out the dev environment setup script in dev/dev-env-setup.sh and make sure you're selecting a Python version >=3.8 (3.9 is preferred on M1's). It (should) work fine. Activate that environment and make sure it's set as the env in your IDE and you should be able to iterate on testing validations a bit faster :) 

Sounds good to me on the updating the functionality in`mlflow.models.utils.py` to support additional functionality for types that you need to enable PyTorch typing. Your tests (and existing) will validate the correct functionality of any changes :) 

Thanks again for picking this up and let me know if you get stuck on getting that dev environment setup!

File: tests/pytorch/test_pytorch_model_export.py [hunk 1]
```diff
@@ -322,16 +322,18 @@ def test_pyfunc_model_works_with_np_input_type(
     assert type(np_result) == np.ndarray
     np.testing.assert_array_almost_equal(np_result[:, 0], sequential_predicted, decimal=4)
 
-    # predict does not work with lists
+    # predict does not work with scalars
     with pytest.raises(
-        TypeError, match="The PyTorch flavor does not support List or Dict input types"
+        TypeError, match="Input data should be pandas.DataFrame, torch.Tensor, or numpy.ndarray"
```
Comment (BenWilson2) [inline tests/pytorch/test_pytorch_model_export.py:327]:
```suggestion
        TypeError, match="Input must be pandas.DataFrame, torch.Tensor, or numpy.ndarray or iterable thereof."
```

File: mlflow/pytorch/__init__.py [hunk 2]
```diff
@@ -747,34 +747,64 @@ def __init__(self, pytorch_model):
...
+        elif isinstance(data, (dict, OrderedDict)):
+            input_tensor = {k: _to_tensor(v) for k, v in data.items()}
+        elif isinstance(data, (pd.DataFrame, np.ndarray, torch.Tensor)):
+            input_tensor = _to_tensor(data)
+        else:
             raise TypeError(
-                "The PyTorch flavor does not support List or Dict input types. "
-                "Please use a pandas.DataFrame or a numpy.ndarray"
+                "Input must be pandas.DataFrame, torch.Tensor, or numpy.ndarray"
+                " or iterable thereof!"
```
Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:789]:
```suggestion
                " or iterable thereof."
```

File: tests/pytorch/test_pytorch_model_export.py (hunk 1 above)
Comment (BenWilson2) [reply tests/pytorch/test_pytorch_model_export.py:327]: The regex match is failing here in the PyTorch cross version test suites

**Comment by dbczumar** [comment]
Hi @IngoStatworx , any updates here?
